File: Search in a 2D matrix II.py
# Time Complexity: O(m+n)
# Space Complexity: O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        if m==0 and n==0:
            return False
        r = m - 1
        c = 0
        while(r>=0 and c <n):
            if target > matrix[r][c]:
                c += 1
            elif target < matrix[r][c]:
                r -= 1
            else:
                return True

File: search_in_a_2D_matrix_II.java
// Time Complexity : O(m + n); m = matrix.length; n = matrix[0].length
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
class Solution {
    private boolean inBounds(int x, int y, int m, int n) {
        return x >= 0 && x < m && y >=0 && y < n;
    }
    
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }

        int m = matrix.length;
        int n = matrix[0].length;
        if(target < matrix[0][0] || target > matrix[m-1][n-1]) {
            return false;
        }

        int curr = 0;
        int i = 0;
        int j = n - 1;

        while(inBounds(i, j, m, n)) {
            curr = matrix[i][j];
            if(target == curr) {
                return true;
            }
            else if(target > curr) {
                // move down
                i += 1;
            } else {
                // move left
                j -= 1;
            }
        }

        return false;
    }
}